### PR TITLE
fix/ local DigiFinex order book not in sync and occassional websocket disconnects

### PR DESCRIPTION
**Pull request by CrimsonJacket**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
The issue is the result of not including the right resolution for `timestamp` values for both `SNAPSHOT` and `DIFF` order book messages. This results in some diffs being dropped by the `OrderBookTracker` 

The issue of the frequent websocket disconnects seems to occur since the APIOrderBookDataSource is handling 2 connections for orderbook diff and orderbook trade updates. Refactoring the logic to utilize 1 websocket client seems to resolve the issue.

Resolves [sc-28591] [sc-28603]

**Tests performed by the developer**:
No unit tests have been added since the Digifinex connector uses an old architecture and that this isnt a connector we supposedly maintained.

**Tips for QA testing**:
Ensure that the orderbook for liquid markets are now in sync.
The frequent ws disconnects should no longer be an issue.

### CODE REVIEW

**Comment by gerald-hb** [comment]
Getting Websocket error. One incident, running the bot for 2 1/2 hours. I will continue to run this overnight and check if there are other WS disconnection will occur.
image
```2022-06-16 12:00:02,511 - 22219 - hummingbot.connector.exchange.digifinex.digifinex_api_order_book_data_source - NETWORK - Unexpected error with WebSocket connection.
Traceback (most recent call last):
  File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_api_order_book_data_source.py", line 191, in listen_for_subscriptions
    await self._subscribe_channels(ws)
  File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_api_order_book_data_source.py", line 172, in _subscribe_channels
    await websocket.subscribe("depth", list(map(
  File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 152, in subscribe
    id = await self.request(category + ".subscribe", channels)
  File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 148, in request
    return await self._emit(method, data)
  File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 142, in _emit
    await self._client.send(req)
AttributeError: 'NoneType' object has no attribute 'send'
2022-06-16 12:00:02,511 - 22219 - hummingbot.connector.exchange.digifinex.digifinex_api_order_book_data_source - WARNING - Unexpected error with WebSocket connection. Retrying in 30 seconds. Check network connection.
2022-06-16 12:00:02,514 - 22219 - hummingbot.connector.exchange.digifinex.digifinex_websocket - ERROR - Websocket error: 'server rejected WebSocket connection: HTTP 502'
Traceback (most recent call last):
  File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 50, in connect
    self._client = await websockets.connect(self._WS_URL)
  File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/site-packages/websockets/legacy/client.py", line 629, in __await_impl__
    await protocol.handshake(
  File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/site-packages/websockets/legacy/client.py", line 388, in handshake
    raise InvalidStatusCode(status_code)
websockets.exceptions.InvalidStatusCode: server rejected WebSocket connection: HTTP 502
2022-06-16 12:00:02,514 - 22219 - hummingbot.connector.exchange.digifinex.digifinex_websocket - NETWORK - {"id":1655380802514410,"method":"depth.subscribe","params":["DOGE_USDT"]}
2022-06-16 12:00:02,514 - 22219 - hummingbot.connector.exchange.digifinex.digifinex_api_order_book_data_source - ERROR - Unexpected error occured subscribing to Digifinex public channels. ('NoneType' object has no attribute 'send')
2022-06-16 12:00:02,514 - 22219 - hummingbot.connector.exchange.digifinex.digifinex_api_order_book_data_source - NETWORK - Unexpected error with WebSocket connection.
Traceback (most recent call last):
  File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_api_order_book_data_source.py", line 191, in listen_for_subscriptions
    await self._subscribe_channels(ws)
  File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_api_order_book_data_source.py", line 172, in _subscribe_channels
    await websocket.subscribe("depth", list(map(
  File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 152, in subscribe
    id = await self.request(category + ".subscribe", channels)
  File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 148, in request
    return await self._emit(method, data)
  File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 142, in _emit
    await self._client.send(req)
AttributeError: 'NoneType' object has no attribute 'send'
```
Config used 
(Test is performed using source build on AWS cloud)
image

**Comment by rxlxrxsx** [comment]
There are still instances that the order book is not in sync with the exchange. See clip attached running on MDT-USDT pair:

https://user-images.githubusercontent.com/64377055/174092952-3e682502-043e-4984-9a5d-53110f3ca95a.mov

**Comment by gerald-hb** [comment]
> Getting Websocket error. One incident, running the bot for 2 1/2 hours. I will continue to run this overnight and check if there are other WS disconnection will occur. image
> 
> ```
> Traceback (most recent call last):
>   File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_api_order_book_data_source.py", line 191, in listen_for_subscriptions
>     await self._subscribe_channels(ws)
>   File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_api_order_book_data_source.py", line 172, in _subscribe_channels
>     await websocket.subscribe("depth", list(map(
>   File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 152, in subscribe
>     id = await self.request(category + ".subscribe", channels)
>   File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 148, in request
>     return await self._emit(method, data)
>   File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 142, in _emit
>     await self._client.send(req)
> AttributeError: 'NoneType' object has no attribute 'send'
> 2022-06-16 12:00:02,511 - 22219 - hummingbot.connector.exchange.digifinex.digifinex_api_order_book_data_source - WARNING - Unexpected error with WebSocket connection. Retrying in 30 seconds. Check network connection.
> 2022-06-16 12:00:02,514 - 22219 - hummingbot.connector.exchange.digifinex.digifinex_websocket - ERROR - Websocket error: 'server rejected WebSocket connection: HTTP 502'
> Traceback (most recent call last):
>   File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 50, in connect
>     self._client = await websockets.connect(self._WS_URL)
>   File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/site-packages/websockets/legacy/client.py", line 629, in __await_impl__
>     await protocol.handshake(
>   File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/site-packages/websockets/legacy/client.py", line 388, in handshake
>     raise InvalidStatusCode(status_code)
> websockets.exceptions.InvalidStatusCode: server rejected WebSocket connection: HTTP 502
> 2022-06-16 12:00:02,514 - 22219 - hummingbot.connector.exchange.digifinex.digifinex_websocket - NETWORK - {"id":1655380802514410,"method":"depth.subscribe","params":["DOGE_USDT"]}
> 2022-06-16 12:00:02,514 - 22219 - hummingbot.connector.exchange.digifinex.digifinex_api_order_book_data_source - ERROR - Unexpected error occured subscribing to Digifinex public channels. ('NoneType' object has no attribute 'send')
> 2022-06-16 12:00:02,514 - 22219 - hummingbot.connector.exchange.digifinex.digifinex_api_order_book_data_source - NETWORK - Unexpected error with WebSocket connection.
> Traceback (most recent call last):
>   File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_api_order_book_data_source.py", line 191, in listen_for_subscriptions
>     await self._subscribe_channels(ws)
>   File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_api_order_book_data_source.py", line 172, in _subscribe_channels
>     await websocket.subscribe("depth", list(map(
>   File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 152, in subscribe
>     id = await self.request(category + ".subscribe", channels)
>   File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 148, in request
>     return await self._emit(method, data)
>   File "/home/ubuntu/pr261/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 142, in _emit
>     await self._client.send(req)
> AttributeError: 'NoneType' object has no attribute 'send'
> ```
> 
> Config used (Test is performed using source build on AWS cloud) image

After the overnight run, we are still seeing multiple Websocket disconnection. Attaching the logs for reference
[logs_conf_pure_mm_1.log](https://github.com/CoinAlpha/hummingbot/files/8923622/logs_conf_pure_mm_1.log)

**Comment by rxlxrxsx** [comment]
- Setup a bot running for 4 hours already and no websocket connection errors so far, will continue to run the bot and update here if any websocket disconnection are encountered.

- On the latest commit the order book is no longer updating after starting the bot. See attached clip running on ETH-USDT pair:

https://user-images.githubusercontent.com/64377055/174244211-4e338957-0124-46eb-9a7a-ea14912c96be.mov

**Comment by Deepali6997** [comment]
getting the following error immediately after running the bot : Attached logs for reference.

Screenshot 2022-06-17 at 3 54 39 PM

[logs_pmm2.log](https://github.com/CoinAlpha/hummingbot/files/8926724/logs_pmm2.log)

**Comment by rxlxrxsx** [comment]
Still having issues with order book synchronization as well on the latest commit:

https://user-images.githubusercontent.com/64377055/174295255-a165e1ad-f29c-41de-8a96-e5a2a67dbf11.mov

**Comment by CrimsonJacket** [comment]
> Still having an issue with order book synchronization as well on the latest commit:
After some discussions with @aarmoa, it is evident that there are some underlying issues with DigiFinex's WS API. Even on the web application we are noticing frequent disconnects. As such, we opted for simply handling those disconnects and reconnecting to the WS API. This, without a doubt, would introduce certain inconsistencies.

Also I'd like to add that at this stage, with an unreliable WS connection and absence of API Rate Limits that are not present the connector should **not** be considered reliable.

**Comment by rxlxrxsx** [comment]
- Getting flooded by websocket connection errors immediately after starting the bot:

Screen Shot 2022-06-20 at 1 43 41 PM

```python
2022-06-20 12:29:30,025 - 3506 - hummingbot.connector.exchange.digifinex.digifinex_api_order_book_data_source - ERROR - Unexpected error with WebSocket connection. Concurrent call to receive() is not allowed
Traceback (most recent call last):
  File "/Users/reiniersalazar/hummingbot1/hummingbot/connector/exchange/digifinex/digifinex_api_order_book_data_source.py", line 186, in listen_for_subscriptions
    await self._ws.subscribe("depth", trading_pairs)
  File "/Users/reiniersalazar/hummingbot1/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 101, in subscribe
    raw_msg: bytes = await self._websocket.receive_bytes()
  File "/Users/reiniersalazar/opt/anaconda3/envs/hummingbot/lib/python3.8/site-packages/aiohttp/client_ws.py", line 280, in receive_bytes
    msg = await self.receive(timeout)
  File "/Users/reiniersalazar/opt/anaconda3/envs/hummingbot/lib/python3.8/site-packages/aiohttp/client_ws.py", line 216, in receive
    raise RuntimeError("Concurrent call to receive() is not allowed")
RuntimeError: Concurrent call to receive() is not allowed
```

```python
2022-06-20 12:29:30,032 - 3506 - hummingbot.connector.exchange.digifinex.digifinex_api_order_book_data_source - ERROR - Unexpected error with WebSocket connection. Received message 256:None is not bytes
Traceback (most recent call last):
  File "/Users/reiniersalazar/hummingbot1/hummingbot/connector/exchange/digifinex/digifinex_api_order_book_data_source.py", line 186, in listen_for_subscriptions
    await self._ws.subscribe("depth", trading_pairs)
  File "/Users/reiniersalazar/hummingbot1/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 101, in subscribe
    raw_msg: bytes = await self._websocket.receive_bytes()
  File "/Users/reiniersalazar/opt/anaconda3/envs/hummingbot/lib/python3.8/site-packages/aiohttp/client_ws.py", line 282, in receive_bytes
    raise TypeError(f"Received message {msg.type}:{msg.data!r} is not bytes")
TypeError: Received message 256:None is not bytes
```

```python
2022-06-20 12:29:30,716 - 3506 - hummingbot.connector.exchange.digifinex.digifinex_api_order_book_data_source - ERROR - Unexpected error with WebSocket connection. local variable 'parse_msg' referenced before assignment
Traceback (most recent call last):
  File "/Users/reiniersalazar/hummingbot1/hummingbot/connector/exchange/digifinex/digifinex_api_order_book_data_source.py", line 189, in listen_for_subscriptions
    async for msg in self._ws.iter_messages():
  File "/Users/reiniersalazar/hummingbot1/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 134, in iter_messages
    yield parse_msg
UnboundLocalError: local variable 'parse_msg' referenced before assignment
```

```python
2022-06-20 12:29:31,007 - 3506 - hummingbot.connector.exchange.digifinex.digifinex_api_order_book_data_source - ERROR - Unexpected error with WebSocket connection. 'NoneType' object has no attribute 'send_str'
Traceback (most recent call last):
  File "/Users/reiniersalazar/hummingbot1/hummingbot/connector/exchange/digifinex/digifinex_api_order_book_data_source.py", line 186, in listen_for_subscriptions
    await self._ws.subscribe("depth", trading_pairs)
  File "/Users/reiniersalazar/hummingbot1/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 100, in subscribe
    request_id = await self.request(category + ".subscribe", channels)
  File "/Users/reiniersalazar/hummingbot1/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 94, in request
    await self._websocket.send_str(ujson.dumps(payload))
AttributeError: 'NoneType' object has no attribute 'send_str'
```

```python
2022-06-20 12:29:32,276 - 3506 - hummingbot.connector.exchange.digifinex.digifinex_websocket - ERROR - Websocket error: '401, message='Invalid response status', url=URL('wss://openapi.digifinex.com/ws/v1/')'
Traceback (most recent call last):
  File "/Users/reiniersalazar/hummingbot1/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 47, in connect
    self._websocket = await aiohttp.ClientSession().ws_connect(self._WS_URL)
  File "/Users/reiniersalazar/opt/anaconda3/envs/hummingbot/lib/python3.8/site-packages/aiohttp/client.py", line 769, in _ws_connect
    raise WSServerHandshakeError(
aiohttp.client_exceptions.WSServerHandshakeError: 401, message='Invalid response status', url=URL('wss://openapi.digifinex.com/ws/v1/')
```

Log file: [logs_conf_pure_mm_1.log](https://github.com/CoinAlpha/hummingbot/files/8937920/logs_conf_pure_mm_1.log)

**Comment by CrimsonJacket** [comment]
@rxlxrxsx Thanks for testing it again.
The 401 error response seems to suggest that the API key is "Not Authorised". Could you try with a new set of API keys?
The next few errors should not keep happen indefinitely and should be handled accordingly. I will look into this and make more changes.

**Comment by Deepali6997** [comment]
Would like to add one more observation. When adding digifinex, we were getting the following error message. {'code': 10008}. When trying once again, the API key got accepted. 

Screenshot 2022-06-20 at 2 32 41 PM
 
Also, this error message is different from the error message that comes when we give wrong API keys
Screenshot 2022-06-20 at 2 34 01 PM

P.S. : Used the existing API keys

**Comment by Deepali6997** [comment]
Getting same errors with new set of API keys 

Screenshot 2022-06-20 at 4 19 37 PM

log file : [logs_PMM.log](https://github.com/CoinAlpha/hummingbot/files/8939992/logs_PMM.log)

**Comment by rxlxrxsx** [comment]
Checking the long term bot running for 14 hours the websocket errors previously reported were no longer appearing. Here are the websocket errors that the bot was able to encounter while currently running on the latest commit:

32 instances of this error

```python
2022-06-20 16:00:03,036 - 24188 - hummingbot.connector.exchange.digifinex.digifinex_websocket - ERROR - Websocket error: '502, message='Invalid response status', url=URL('wss://openapi.digifinex.com/ws/v1/')'
Traceback (most recent call last):
  File "/home/ubuntu/hummingbot/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 43, in connect
    self._connection = await self._client.ws_connect(self._WS_URL)
  File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/site-packages/aiohttp/client.py", line 769, in _ws_connect
    raise WSServerHandshakeError(
aiohttp.client_exceptions.WSServerHandshakeError: 502, message='Invalid response status', url=URL('wss://openapi.digifinex.com/ws/v1/')
```

5 instances of this error
```python
2022-06-20 16:05:23,887 - 24188 - hummingbot.connector.exchange.digifinex.digifinex_websocket - ERROR - Websocket error: 'Websocket server closed the connection'
Traceback (most recent call last):
  File "/home/ubuntu/hummingbot/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 45, in connect
    await self.login()
  File "/home/ubuntu/hummingbot/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 52, in login
    response: aiohttp.WSMessage = await self.receive()
  File "/home/ubuntu/hummingbot/hummingbot/connector/exchange/digifinex/digifinex_websocket.py", line 107, in receive
    raise aiohttp.ClientConnectionError("Websocket server closed the connection")
aiohttp.client_exceptions.ClientConnectionError: Websocket server closed the connection
```

And occassionally getting this in the logs as well
Screen Shot 2022-06-21 at 12 06 08 PM

Log file: [logs_conf_pure_mm_1.log](https://github.com/CoinAlpha/hummingbot/files/8945796/logs_conf_pure_mm_1.log)

**Comment by rxlxrxsx** [comment]
> After some discussions with @aarmoa, it is evident that there are some underlying issues with DigiFinex's WS API. Even on the web application we are noticing frequent disconnects. As such, we opted for simply handling those disconnects and reconnecting to the WS API. This, without a doubt, would introduce certain inconsistencies.

> Also I'd like to add that at this stage, with an unreliable WS connection and absence of API Rate Limits that are not present the connector should not be considered reliable.

After some discussions with @CrimsonJacket, due to the underlying issue with DigiFinex API we should label the connector as **RED**.

**Comment by CrimsonJacket** [comment]
@rxlxrxsx Yea the above errors above are expected. As per discussed with @aarmoa, the underlying issues are attributed to DigiFinex WS API prematurely disconnecting the websocket connection, thus the errors you have observed. The changes I've made revolves around handling the frequent disconnects and ensuring that the bot resumes.

As for now, it is critical the bot is still able to update the orderbook(relatively in sync), placing orders and tracking orders.
At this stage, I would lean towards labelling the DigiFinex connector as `RED`, what do you think?

**Review by rxlxrxsx — APPROVED** [review]
Tested and confirmed working by QA

- Connected API keys (valid and invalid key) and confirmed working as expected
- Confirmed wallet and exchange balances are reflecting correct
- Created a pure_mm strategy and went through initial prompts successfully
- Started the strategy without issues, confirmed orders are created and cancelled based on config
- Confirmed order book prices is relatively in sync with the exchange
- Trade information are correct (price, size and maker/taker fees) based on history command output
- Compared trades in client and exchange, no discrepancies found
- Confirmed hanging order/cancellation and multiple order levels works as expected

File: hummingbot/connector/exchange/digifinex/digifinex_api_user_stream_data_source.py [hunk 1]
```diff
@@ -102,12 +50,31 @@ async def listen_for_user_stream(self, output: asyncio.Queue):
...
+                self.logger().warning("Attemping re-connection with Websocket Private Channels...")
+            except Exception as e:
                 self.logger().error(
-                    "Unexpected error with Digifinex WebSocket connection. " "Retrying after 30 seconds...", exc_info=True
+                    f"Unexpected error with WebSocket connection. {str(e)}",
+                    exc_info=True
                 )
-                await asyncio.sleep(30.0)
+            finally:
+                await self._ws.disconnect()
```
Comment (aarmoa) [inline hummingbot/connector/exchange/digifinex/digifinex_api_user_stream_data_source.py:80]: Here we should add the same check that self._ws is not None, as in the order book data source

**Review by aarmoa — CHANGES_REQUESTED** [review]

**Review by aarmoa — APPROVED** [review]
The changes look good to me

**Comment by CrimsonJacket** [comment]
Closed and recreated [PR #5430](https://github.com/hummingbot/hummingbot/pull/5430) created on Foundation repo.
Snapshot Link: https://snapshot.org/#/hbot-prp.eth/proposal/0xd49b9ba5c2c34aa33592732d7f59e6da45f41ec519774e78e0dc1c4ef25d5d2a

**Pull request closed without merging**